File: Lucy/src/2025-8-W01/17_letter_combinations_of_a_phone_number.ts
function letterCombinations(digits: string): string[] {
  if (digits.length === 0) return [];

  const results: string[] = [];

  const digitMap = new Map<string, string>();
  const letters = ["abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"];
  for (let i = 0; i < letters.length; i++) {
    digitMap.set((i + 2).toString(), letters[i]);
  }

  function backtrack(letters: string, currentIndex: number) {
    if (letters.length === digits.length) {
      results.push(letters);
      return;
    }

    const currentNumber = digits[currentIndex];

    for (const ch of digitMap.get(currentNumber) || "") {
      backtrack(letters + ch, currentIndex + 1);
    }
  }

  backtrack("", 0);

  return results;
}

File: Seori/17. Letter Combinations of a Phone Number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        letters = ['', '', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
        answer = []
        def backtracking(index, arr, string):
            if index == len(arr):
                answer.append(string)
                return

            for letter in letters[arr[index]]:
                backtracking(index+1, arr, string+letter)
            
        if digits == '': return []
        arr = [int(num) for num in digits]
        n = len(arr)
        backtracking(0, arr, '')
        return answer
